xds/csds/csds_test.go

```diff
@@ -0,0 +1,672 @@
+				wantLis = append(wantLis, &v3adminpb.ListenersConfigDump_DynamicListener{
+					Name: ldsTargets[i],
+					ActiveState: &v3adminpb.ListenersConfigDump_DynamicListenerState{
+						VersionInfo: wantVersion,
+						Listener:    listenerAnys[i],
+						LastUpdated: nil,
+					},
+					ErrorState:   nil,
+					ClientStatus: v3adminpb.ClientResourceStatus_ACKED,
+				})
+			}
+			wantDump := &v3adminpb.ListenersConfigDump{
+				VersionInfo:      wantVersion,
+				DynamicListeners: wantLis,
+			}
+			if diff := cmp.Diff(config.ListenerConfig, wantDump, cmpOpts); diff != "" {
+				return fmt.Errorf(diff)
+			}
+		case *v3statuspb.PerXdsConfig_RouteConfig:
+			var wantRoutes []*v3adminpb.RoutesConfigDump_DynamicRouteConfig
+			for i := range rdsTargets {
+				wantRoutes = append(wantRoutes, &v3adminpb.RoutesConfigDump_DynamicRouteConfig{
+					VersionInfo:  wantVersion,
+					RouteConfig:  routeAnys[i],
+					LastUpdated:  nil,
+					ClientStatus: v3adminpb.ClientResourceStatus_ACKED,
+				})
+			}
+			wantDump := &v3adminpb.RoutesConfigDump{
+				DynamicRouteConfigs: wantRoutes,
+			}
+			if diff := cmp.Diff(config.RouteConfig, wantDump, cmpOpts); diff != "" {
+				return fmt.Errorf(diff)
+			}
+		case *v3statuspb.PerXdsConfig_ClusterConfig:
+			var wantCluster []*v3adminpb.ClustersConfigDump_DynamicCluster
+			for i := range cdsTargets {
+				wantCluster = append(wantCluster, &v3adminpb.ClustersConfigDump_DynamicCluster{
+					VersionInfo:  wantVersion,
+					Cluster:      clusterAnys[i],
+					LastUpdated:  nil,
+					ClientStatus: v3adminpb.ClientResourceStatus_ACKED,
+				})
+			}
+			wantDump := &v3adminpb.ClustersConfigDump{
+				VersionInfo:           wantVersion,
+				DynamicActiveClusters: wantCluster,
+			}
+			if diff := cmp.Diff(config.ClusterConfig, wantDump, cmpOpts); diff != "" {
+				return fmt.Errorf(diff)
+			}
+		case *v3statuspb.PerXdsConfig_EndpointConfig:
+			var wantEndpoint []*v3adminpb.EndpointsConfigDump_DynamicEndpointConfig
+			for i := range cdsTargets {
+				wantEndpoint = append(wantEndpoint, &v3adminpb.EndpointsConfigDump_DynamicEndpointConfig{
+					VersionInfo:    wantVersion,
+					EndpointConfig: endpointAnys[i],
+					LastUpdated:    nil,
+					ClientStatus:   v3adminpb.ClientResourceStatus_ACKED,
+				})
+			}
+			wantDump := &v3adminpb.EndpointsConfigDump{
+				DynamicEndpointConfigs: wantEndpoint,
+			}
+			if diff := cmp.Diff(config.EndpointConfig, wantDump, cmpOpts); diff != "" {
+				return fmt.Errorf(diff)
+			}
+		default:
+			return fmt.Errorf("unexpected PerXdsConfig: %+v; %v", cfg.PerXdsConfig, protoToJSON(r))
+		}
+	}
+	return nil
+}
+
+func checkForNACKed(nackResourceIdx int, stream v3statuspb.ClientStatusDiscoveryService_StreamClientStatusClient) error {
+	const (
+		ackVersion  = "1"
+		nackVersion = "2"
+	)
+
+	if err := stream.Send(&v3statuspb.ClientStatusRequest{Node: nil}); err != nil {
+		return fmt.Errorf("failed to send: %v", err)
+	}
+	r, err := stream.Recv()
```

Comment:
We are using the streaming method in all tests. For completeness sake, should one of the `checkForXxx` use the unary method?

Reply:
But isn't that an implementation detail. Hmm ... maybe its ok.